File: backend/app/models/book.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import TYPE_CHECKING

from sqlalchemy import DateTime, Index, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from backend.app.database.session import Base

if TYPE_CHECKING:
    from backend.app.models.borrow import BorrowRecord


def utcnow() -> datetime:
    return datetime.now(timezone.utc)


class Book(Base):
    __tablename__ = "books"
    __table_args__ = (
        Index("ix_books_title", "title"),
        Index("ix_books_author", "author"),
        Index("ix_books_category", "category"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    author: Mapped[str] = mapped_column(String(100), nullable=False)
    isbn: Mapped[str | None] = mapped_column(String(20), unique=True, nullable=True)
    publisher: Mapped[str | None] = mapped_column(String(200), nullable=True)
    publish_year: Mapped[int | None] = mapped_column(Integer, nullable=True)
    category: Mapped[str | None] = mapped_column(String(50), nullable=True)
    total_copies: Mapped[int] = mapped_column(Integer, default=1, nullable=False)
    available_copies: Mapped[int] = mapped_column(Integer, default=1, nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow, nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utcnow, onupdate=utcnow, nullable=False
    )

    borrow_records: Mapped[list["BorrowRecord"]] = relationship(
        "BorrowRecord", back_populates="book", passive_deletes=True
    )

File: frontend/src/types/book.ts
// 图书类型，契约见 docs/system-design.md 8.4
export interface BookOut {
  id: number
  title: string
  author: string
  isbn: string | null
  publisher: string | null
  publish_year: number | null
  category: string | null
  total_copies: number
  available_copies: number
  description: string | null
  created_at: string
  updated_at: string
}

export interface BookCreate {
  title: string
  author: string
  isbn?: string
  publisher?: string
  publish_year?: number
  category?: string
  total_copies?: number
  description?: string
}

export type BookUpdate = Partial<BookCreate>

export interface BookBriefOut {
  id: number
  title: string
  author: string
  isbn: string | null
}

export interface BookQuery {
  page?: number
  page_size?: number
  keyword?: string
  category?: string
  available_only?: boolean
}
